### ISSUE TITLE
Globale Funktionen in Blockly

### ISSUE THREAD

**Issue by ChrMaass** [ioBroker/ioBroker.javascript#1312]
Es wäre super, wenn es bei Funktionsblöcken in Blockly die Möglichkeit gäbe, diese als "Global" zu definieren und diese dann als Blöcke zum Aufrufen in jedem Script zur Verfügung zu haben. Der Umweg über einen zusätzlichen Funktionsblock, in welchen dann eine andere Funktion aufgerufen hat, ist unschön und sieht unsauber aus, vor allem wenn man die Funktion dann umbenennt.

**Comment by klein0r** [comment]
Das ist technisch leider nicht so einfach. Denn dann müsste man auch verhindern, dass diese Funktionen jemals umbenannt oder die Global-Scripts gelöscht werden können, falls diese noch irgendwo in Verwendung sind. Mal eben einen Parameter hinzufügen geht dann ja nicht.

"Global" heißt in diesem Fall ja einfach, dass das Script vor jedes andere kopiert wird. Als würde man den Code nehmen und vor jedes andere Script kopieren. Daher wird ja auch die komplette Instanz neugestartet, sobald sich ein global script ändert.

**Comment by ldittmar81** [comment]
Ja klar... Die Skripte werden auch schon jetzt davor rein geschrieben und auch jetzt kann man schon von dort Skripte löschen, auch wenn diese irgendwo gebraucht werden. Auch mit dem Export würde es nicht mehr so funktionieren bzw müsste man auch anpassen, aber machbar wäre es.

Hmm.... Werden die globale Skripte auch jetzt schon vor dem Blockly Skript geschrieben?

**Comment by klein0r** [comment]
Ja, es wird einfach das generierte JavaScript davor kopiert.

**Comment by ldittmar81** [comment]
@ChrMaass Soo... ich habe das ganze mal durchgespielt und es ist jetzt schon möglich globale Skripte in Blockly zu nutzen. Ich habe, als Beispiel, ein Blockly unter global erzeugt:

<img width="1264" alt="Bild1" src="https://github.com/ioBroker/ioBroker.javascript/assets/9552828/e8f8ee55-2b95-4a40-8861-6f632aa6ed0b">

Jetzt steht diese Funktion für alle Skripte zur Verfügung... auch Blockly! Dann nehme ich aus Funktionen ein "JS-Funktion mit Ergebis"-Block und schreibe dort lediglich ein return mit der Funktion aus dem global Ordner:

<img width="402" alt="Bild2" src="https://github.com/ioBroker/ioBroker.javascript/assets/9552828/e0e480d6-33fa-4150-9886-c8501e6c19c6">

<img width="916" alt="Bild3" src="https://github.com/ioBroker/ioBroker.javascript/assets/9552828/44387bb6-5664-438d-a4ad-d156e14eca18">

@klein0r Die globale Funktionen werden ja auch im normalen JS auch vorgeschlagen, wenn man mit dem Tippen anfängt. Man könnte vielleicht im Blockly die Funktionen aus dem global Ordner als Block anbieten. Naja... da das Ganze jetzt auch schon nutzbar ist, hätte der Ticket keine hohe Prio.

**Comment by Diginix** [comment]
Genau so mache ich das auch schon lange. Ich habe ein globales Blockly 
![image](https://github.com/ioBroker/ioBroker.javascript/assets/39678667/a8631c80-5790-43a2-9810-f7f0ee8e789a)
und in allen anderen Blockly gibt es
![image](https://github.com/ioBroker/ioBroker.javascript/assets/39678667/f4643ea1-5531-48c5-9e20-3e2062e9b6b9)
mit
`alexavoice(ausgabe,echos,ssml,speak);`
Dadurch kann ich dann überall in den Blocklly den Block verwenden:
![image](https://github.com/ioBroker/ioBroker.javascript/assets/39678667/48bd5228-9fbb-47fa-8070-351dc506a546)

**Comment by klein0r** [comment]
> Soo... ich habe das ganze mal durchgespielt und es ist jetzt schon möglich globale Skripte in Blockly zu nutzen.

Ja, logischerweise geht das so. Und jetzt stop mal das globale Script oder benenn die Funktion um. Das müsste der Adapter ja alles abfangen. Das macht es kompliziert. Der Weg "vorwärts" wäre umsetzbar. Dann erwarten die Leute wahrscheinlich auch, dass die Funktion in allen Scripts umbenannt wird (was aktuell technisch gar nicht möglich ist, da das JavaScript aus Blockly auf Client-Seite generiert wird).

**Comment by ldittmar81** [comment]
Naja... Wenn man z.B. ein Adapter entfernt (z.B. Telegram oder Awtrix) und irgendwo noch den ein Block des Adapters nutzt.... oder selbst im Javascript die Funktion integriert, dann geht das ganz auch nicht und man bekommt Fehlermeldungen.

Und wenn man globale Skripte in Javascript nutzt und die Skripte dann umbenennt, hat man auch jetzt die Probleme. Wenn man Globale Skripte nutzt, muss man wissen was man tut. Deswegen ist der Ordner auch nur sichtbar, wenn man auf Expertenmodus umschaltet.

Aber warte mal... wenn man eine Funktion im Blockly umbenennt, dann wird das im ganzen Skript umbenannt. Die Blöcke werden auch immer automatisch, aus dem globalen Skript, generiert.... dann würde doch das ganze auch "automatisch" umbenannt werden.... Wenn im Blockly Blöcke Änderungen gemacht werden, greifen diese auch auf alle Blöcke in Nutzung.

**Comment by ChrMaass** [comment]
Was haltet ihr denn davon, wenn in Blockly als Name nur eine art Alias angezeigt wird, den man frei vergeben (und ändern kann), aber intern die als global Markierte Blockly-Funktion einen generierten Namen erhält, den man nicht ändern kann.

Parameter sind doch in JS in der Regel ehr optional, so das es selbst bei Änderung der Parameter nicht zu großen Problemen kommen sollte, oder? Ansonsten friert man die Parameter automatisch ein, sobald mal ein Script auf Global setzt und muss zur Änderung der Parameter eine neue Funktion machen (und die Blöcke rüber kopieren).

Wenn man das schafft, es so zu definieren, dann könnte man Blockly in Zukunft frei erweitern und quasi Blockvorlagen teilen. Alleine durch so kleine Helferfunktionen wie: Setze den Wert nur auf folgenden Wert, wenn der größer als der aktuelle Wert ist (Rollos z.B.) usw., die man immer wieder mal benötigt.

Freue mich über euer Feedback

**Comment by ChrMaass** [comment]
Hallo zusammen,

ich wollte das Thema nochmal aufgreifen, da ich gerade erneut in eine Situation gekommen bin, in der ein erweiterter Return-Block in Blockly extrem hilfreich gewesen wäre.

Habt ihr euch zwischenzeitlich Gedanken zur Umsetzung gemacht?

Wenn man die Idee etwas weiter denkt, könnte man damit den Grundstein für eine Art Script-Vorlagen-Manager legen. So könnten erprobte und wiederverwendbare Skriptbausteine einfach geteilt und eingebunden werden – ohne die Einstiegshürde eines eigenen Adapters. Blockly würde dadurch deutlich flexibler und zugänglicher, insbesondere für Nutzer, die schnell eigene Funktionen (z. B. zum Updaten von DNS-Einträgen bei Hetzner) erstellen und mit anderen teilen möchten.

Das würde Blockly langfristig nicht nur funktional erweitern, sondern auch die Community stärker einbinden und fördern.

Was haltet ihr davon?
